actors/miner/src/expiration_queue.rs

```diff
@@ -363,64 +364,50 @@ impl<'db, BS: Blockstore> ExpirationQueue<'db, BS> {
         // We expect this to find all recovering sectors within the first FaultMaxAge/WPoStProvingPeriod entries
         // (i.e. 14 for 14-day faults), but if something has gone wrong it's safer not to fail if that's not met.
         let mut sectors_rescheduled = Vec::<&SectorOnChainInfo>::new();
         let mut recovered_power = PowerPair::zero();
+        let mut sectors_by_number = BTreeMap::<u64, &SectorOnChainInfo>::new();
 
-        self.iter_while_mut(|_epoch, expiration_set| {
-            let on_time_sectors: BTreeSet<SectorNumber> = expiration_set
-                .on_time_sectors
-                .bounded_iter(ENTRY_SECTORS_MAX)
-                .context("too many sectors to reschedule")?
-                .map(|i| i as SectorNumber)
-                .collect();
-
-            let early_sectors: BTreeSet<SectorNumber> = expiration_set
-                .early_sectors
-                .bounded_iter(ENTRY_SECTORS_MAX)
-                .context("too many sectors to reschedule")?
-                .map(|i| i as SectorNumber)
-                .collect();
-
-            // This loop could alternatively be done by constructing bitfields and intersecting them, but it's not
-            // clear that would be much faster (O(max(N, M)) vs O(N+M)).
-            // If faults are correlated, the first queue entry likely has them all anyway.
-            // The length of sectors has a maximum of one partition size.
-            for sector in sectors.iter() {
-                let sector_number = sector.sector_number;
-                let power = power_for_sector(sector_size, sector);
-                let mut found = false;
-
-                if on_time_sectors.contains(&sector_number) {
-                    found = true;
-                    // If the sector expires on-time at this epoch, leave it here but change faulty power to active.
-                    // The pledge is already part of the on-time pledge at this entry.
-                    expiration_set.faulty_power -= &power;
-                    expiration_set.active_power += &power;
-                } else if early_sectors.contains(&sector_number) {
-                    found = true;
-                    // If the sector expires early at this epoch, remove it for re-scheduling.
-                    // It's not part of the on-time pledge number here.
-                    expiration_set.early_sectors.unset(sector_number);
-                    expiration_set.faulty_power -= &power;
-                    sectors_rescheduled.push(sector);
-                }
-
-                if found {
-                    recovered_power += &power;
-                    remaining.remove(&sector.sector_number);
-                }
+        for sector in sectors.iter() {
+            let epoch = self.quant.quantize_up(sector.expiration);
+            sectors_group.entry(epoch).or_insert(BitField::new()).set(sector.sector_number);
```

Comment:
So, this'll store each bit in a btreemap internally, making the intersection logic below a bit slower (maybe)?

I'd try:

1. Pushing the sector numbers into a series of separate `Vec`s.
2. Calling `BitField::try_from_iter(...)` on each of these vecs afterwards.

Reply:
yes, let early_sectors = BitField::try_from_bits(sectors_total)?;  this change is very usefull.
I test the bafy2bzacedz5jt3rllzlxgbpzfqujj52g5fu3hb3soosa6wodj6tujzrvyltq message, total gas is 
2022-09-08T22:57:55.197 ERROR fvm::kernel::default > AL: xxxx  p4; gas: 1447431141.500, origin gas is 1549095894